<a href="https://colab.research.google.com/github/lustraka/Data_Analysis_Workouts/blob/main/Analyse_Twitter_Data/wrangle_act.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project: Wrangling and Analyze Data

In [1]:
# Import dependencies
import requests
import os
import json
import tweepy
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [ ]:
path_csv = 'https://raw.githubusercontent.com/lustraka/Data_Analysis_Workouts/main/Analyse_Twitter_Data/'
dfa = pd.read_csv(path_csv+'twitter-archive-enhanced.csv')
dfa.head()

tweet_id  in_reply_to_status_id  ...  pupper puppo
0  892420643555336193                    NaN  ...    None  None
1  892177421306343426                    NaN  ...    None  None
2  891815181378084864                    NaN  ...    None  None
3  891689557279858688                    NaN  ...    None  None
4  891327558926688256                    NaN  ...    None  None

[5 rows x 17 columns]

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [ ]:
url_tsv = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url_tsv)
with open('image-predictions.tsv', 'wb') as file:
  file.write(r.content)
dfi = pd.read_csv('image-predictions.tsv', sep='\t')
dfi.head()

tweet_id  ... p3_dog
0  666020888022790149  ...   True
1  666029285002620928  ...   True
2  666033412701032449  ...   True
3  666044226329800704  ...   True
4  666049248165822465  ...   True

[5 rows x 12 columns]

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [ ]:
consumer_key = ''
consumer_secret = ''
access_token = ''
access_secret = ''

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [ ]:
from timeit import default_timer as timer

count = 0
fails_dict =  {}
start = timer()

if 'tweet_json.txt' in os.listdir():
  os.remove('tweet_json.txt')

with open('tweet_json.txt', 'a') as file:
  for tweet_id in dfa.tweet_id.values:
    count += 1
    print(str(count) + ': ' + str(tweet_id))
    try:
      status = api.get_status(tweet_id, tweet_mode='extended')._json
      print("Success")
      file.write(json.dumps(status, ensure_ascii=False)+'\n')
    except tweepy.TweepError as e:
      print('Fail')
      fails_dict[tweet_id] = e
      pass
    except e:
      print('Fail', e)
end = timer()
print(f'Elapsed time: {end - start}')
print(fails_dict)

Data gathered form Twitter API:

| Attribute | Type | Description |
| --- | :-: | --- |
| id | int | The integer representation of unique identifier for this Tweet |
| retweet_count | int | Number of times this Tweet has been retweeted. |
| favorite_count | int | *Nullable*. Indicates approximately how many times this tweet has been liked by Twitter users. |

Reference: [Tweepy docs: Tweet Object](https://developer.twitter.com/en/docs/twitter-api/v1/data-dictionary/object-model/tweet)



In [ ]:
df_tweets = []
with open('tweet_json.txt', 'r') as file:
  line = file.readline()
  while line:
    status = json.loads(line)
    df_tweets.append({'tweet_id': status['id'], 'retweet_count': status['retweet_count'], 'favorite_count': status['favorite_count']})
    line = file.readline()
dft = pd.DataFrame(df_tweets)
dft.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7164           34458
1  892177421306343426           5385           29870
2  891815181378084864           3551           22490
3  891689557279858688           7383           37671
4  891327558926688256           7922           35988

In [ ]:
# Store dataframes for further processing if necessary


In [2]:
# Upload the database
url_db = 'https://github.com/lustraka/Data_Analysis_Workouts/blob/main/Analyse_Twitter_Data/weratedogsdata.db?raw=true'
r = requests.get(url_db)
with open('weratedogsdata.db', 'wb') as file:
  file.write(r.content)

from sqlalchemy import create_engine
# Create SQLAlchemy engine and connect to the database
engine = create_engine('sqlite:///weratedogsdata.db')

# Read dataframes from SQlite database
dfa = pd.read_sql('SELECT * FROM dba', engine)
dfi = pd.read_sql('SELECT * FROM dbi', engine)
dft = pd.read_sql('SELECT * FROM dbt', engine)

In [3]:
dfa.shape, dfi.shape, dft.shape

((2356, 17), (2075, 12), (2327, 3))

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



### The archive `twitter_archive_enhanced.csv` (alias `dba`)
> "I extracted this data programmatically, but I didn't do a very good job. The ratings probably aren't all correct. Same goes for the dog names and probably dog stages (see below for more information on these) too. You'll need to assess and clean these columns if you want to use them for analysis and visualization."

In [ ]:
dfa.sample(15)

tweet_id  in_reply_to_status_id  ...  pupper puppo
2174  669216679721873412                    NaN  ...    None  None
326   833826103416520705                    NaN  ...    None  None
2272  667491009379606528                    NaN  ...    None  None
201   853760880890318849                    NaN  ...    None  None
410   823322678127919110                    NaN  ...    None  None
1709  680609293079592961                    NaN  ...    None  None
75    878281511006478336                    NaN  ...    None  None
1815  676613908052996102                    NaN  ...    None  None
247   845459076796616705                    NaN  ...    None  None
473   816336735214911488                    NaN  ...    None  None
610   797236660651966464                    NaN  ...    None  None
1748  679062614270468097                    NaN  ...    None  None
1056  742161199639494656                    NaN  ...    None  None
1052  742528092657332225                    NaN  ...    None  None
1540  689659372465688576                    NaN  ...    None  None

[15 rows x 17 columns]

In [ ]:
dfa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [ ]:
for col in dfa.columns[[10,11,13,14,15,16]]:
  print(dfa[col].unique())

[  13   12   14    5   17   11   10  420  666    6   15  182  960    0
   75    7   84    9   24    8    1   27    3    4  165 1776  204   50
   99   80   45   60   44  143  121   20   26    2  144   88]
[ 10   0  15  70   7  11 150 170  20  50  90  80  40 130 110  16 120   2]
['None' 'doggo']
['None' 'floofer']
['None' 'pupper']
['None' 'puppo']


#### Curated `twitter_archive_enhanced.csv` info

| # | Variable | Non-Null | Nunique | Dtype | Notes |
|---|----------|----------|---------|-------|-------|
| 0 | tweet_id | 2356 | 2356 | int64  | |
| 1 | in_reply_to_status_id | 78 | 77 | float64 | these tweets are replies |
| 2 | in_reply_to_user_id | 78 | 31 | float64 | see $\uparrow$ |
| 3 | timestamp | 2356 | 2356 | object | object $\to$ datetime | 
| 4 | source | 2356 | 4 | object | |
| 5 | text | 2356 | 2356 | object | some tweets don't have an image (1) |
| 6 | retweeted_status_id | 181 | 181 | float64 | these are retweets |
| 7 | retweeted_status_user_id | 181 | 25 | float64 | see $\uparrow$ |
| 8 | retweeted_status_timestamp | 181 | 181 | object | see $\uparrow$ |
| 9 | expanded_urls | 2297 | 2218 | object | missing values |
| 10 | rating_numerator | 2356 | 40 | int64  | entries with numerator $> 20$ may be incorrect (4a) |
| 11 | rating_denominator | 2356 | 18 | int64  | entries with denominator $\neq 10$ may be incorrect (4b) |
| 12 | name | 2356 | 957 | object | incorrect names or missing values (2) |
| 13 | doggo | 2356 | 2 | object | a value as a column + (3) some misclassified stages|
| 14 | floofer | 2356 | 2 | object | see $\uparrow$ |
| 15 | pupper | 2356 | 2 | object | see $\uparrow$ |
| 16 | puppo | 2356 | 2 | object | see $\uparrow$ |

Source: visual and programmatic assessment

```python
# #, Variable, Non-Null (Count), Dtype:
dfa.info()
# Nunique:
dfa.nunique()
# Check unique values
for col in dfa.columns[[10,11,13,14,15,16]]:
  print(dfa[col].unique())
# Notes
# (1) Some tweets don't have an image
dfa.loc[dfa.text.apply(lambda s: 'https://t.co' not in s)].shape[0]
# [Out] 124
```

In [56]:
# (2a) Incorrect names - begin with a lowercase
import re
print(re.findall(r';([a-z].*?);', ';'.join(dfa.name.unique())))

['such', 'a', 'quite', 'not', 'one', 'incredibly', 'mad', 'an', 'very', 'just', 'my', 'his', 'actually', 'getting', 'this', 'all', 'old', 'infuriating', 'the', 'by', 'officially', 'life', 'light', 'space']


In [54]:
# (2b) Incorrect names - None
dfa.loc[dfa.name == 'None'].shape[0]

745

In [26]:
# (3a) Misclassified stages - indicated in the stage but not present in the text
stages = ['doggo', 'pupper', 'puppo', 'floofer']
print('Stage     | Total | Misclassified |')
print('-'*35)
for stage in stages:
  total = dfa.loc[dfa[stage] == stage].shape[0]
  missed = dfa.loc[(dfa[stage] == stage) & (dfa.text.apply(lambda s: stage not in s.lower()))].shape[0]
  print(f"{stage.ljust(9)} | {total:5d} | {missed:13d} |")

Stage     | Total | Misclassified |
-----------------------------------
doggo     |    97 |             0 |
pupper    |   257 |             0 |
puppo     |    30 |             0 |
floofer   |    10 |             0 |


In [27]:
# (3b) Misclassified stages - not indicated in the stage but is present in the text
stages = ['doggo', 'pupper', 'puppo', 'floofer']
print('Stage     | Total | Misclassified |')
print('-'*35)
for stage in stages:
  total = dfa.loc[dfa[stage] == stage].shape[0]
  missed = dfa.loc[(dfa[stage] != stage) & (dfa.text.apply(lambda s: stage in s.lower()))].shape[0]
  print(f"{stage.ljust(9)} | {total:5d} | {missed:13d} |")

Stage     | Total | Misclassified |
-----------------------------------
doggo     |    97 |            10 |
pupper    |   257 |            26 |
puppo     |    30 |             8 |
floofer   |    10 |             0 |


##### Note (4) Ratings where `rating_numerator` $ > 20$ or  `rating_denomiator` $\neq 10$
Code used:
```python
# Show the whole text
pd.options.display.max_colwidth = None

# (4a) Show tweets with possibly incorrect rating : rating_numerator > 20
dfa.loc[dfa.rating_numerator > 20, ['text', 'rating_numerator', 'rating_denominator']]

# (4b) Show tweets with possibly incorrect rating : rating_denominator != 10
dfa.loc[dfa.rating_denominator != 10, ['text', 'rating_numerator', 'rating_denominator']]

```
In cases where users used float numbers, such as 9.75/10 or 11.27/10, we will use the floor rounding, i.e. 9/10 or 11/10 respectively. We will correct only those rating which were incorrectly identified in the text. Ratings with weird values used in the text are left unchanged cos they're good dogs Brent.

Results:

In [9]:
# Show the whole text
pd.options.display.max_colwidth = None

# Fill dict with key = index and value = correct rating
incorrect_rating = {313 : '13/10', 340 : '9/10', 763: '11/10', 1717 : '11/10', 313 : '13/10', 784 : '14/10', 1165 : '13/10', 1202 : '11/10', 1662 : '10/10', 2335 : '9/10'}

# Indicate tweets with missing rating
missing_rating = [342, 516]

# Show tweet with incorrectly identified rating
dfa.loc[list(incorrect_rating.keys()), ['text', 'rating_numerator', 'rating_denominator']]

text  ...  rating_denominator
313                                                                  @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho  ...                   0
340                        RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…  ...                  10
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq  ...                  10
1717                                                                                                     Here's a sleepy Christmas pupper 11/10 https://t.co/KXg0f8GNQ9  ...                  10
784                        RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…  ...                  11
1165                                                                                                   Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a  ...                  20
1202                                              This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq  ...                  50
1662                        This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5  ...                  11
2335                           This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv  ...                   2

[9 rows x 3 columns]

### The Tweet Image Predictions `image_predictions.tsv`
>A table full of image predictions (the top three only) alongside each tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images)

In [57]:
dfi.sample(10)

tweet_id  ... p3_dog
131   668297328638447616  ...      1
1548  792883833364439040  ...      1
1286  750868782890057730  ...      0
709   685268753634967552  ...      1
1999  875747767867523072  ...      1
1723  819924195358416896  ...      0
119   668171859951755264  ...      0
1746  823269594223824897  ...      0
1716  819006400881917954  ...      0
465   675006312288268288  ...      1

[10 rows x 12 columns]

In [58]:
dfi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   int64  
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   int64  
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 194.7+ KB


#### Curated Info

| # | Variable | Non-Null | Nunique | Dtype | Notes |
|---|----------|----------|---------|-------|-------|
| 0 | tweet_id | 2075 | 2078 | int64 | |
| 1 | jpg_url | 2075 | 2009 | object | |
| 2 | img_num | 2075 | 4 | int64 | the image number that corresponded to the most confident prediction|
| 3 | p1 | 2075 | 378 | object | prediction |
| 4 | p1_conf | 2075 | 2006 | float64 | confidence of prediction |
| 5 | p1_dog | 2075 | 2 | int64 | Is the prediction a breed of dog? : int $\to$ bool |
| 6 | p2 | 2075 | 405 | object | dtto |
| 7 | p2_conf | 2075 | 2004 | float64 | dtto |
| 8 | p2_dog | 2075 | 2 | int64 | dtto |
| 9 | p3 | 2075 | 408 | object | dtto |
| 10 | p3_conf | 2075 | 2006 | float64 | dtto |
| 11 | p3_dog | 2075 | 2 | int64 | dtto |

Source: visual and programmatic assessment

```python
# #, Variable, Non-Null (Count), Dtype:
dfa.info()
# Nunique:
dfa.nunique()
```


In [68]:
dfi.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

### Additional Data From Twitter API

In [69]:
dft.sample(10)

tweet_id  retweet_count  favorite_count
1453  693280720173801472           1160            3159
890   756939218950160384           1896            6347
500   810657578271330305           2522           10635
1795  676496375194980353            483            1414
837   763103485927849985           2176            7062
1948  672988786805112832            262             895
1441  694183373896572928            866            2780
2008  671561002136281088           6492           11654
1693  680130881361686529            877            2157
1410  696900204696625153            964            2984

In [70]:
dft.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2327 entries, 0 to 2326
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2327 non-null   int64
 1   retweet_count   2327 non-null   int64
 2   favorite_count  2327 non-null   int64
dtypes: int64(3)
memory usage: 54.7 KB


#### Curated Info

| # | Variable | Non-Null | Nunique | Dtype | Notes |
|---|----------|----------|---------|-------|-------|
| 0 | tweet_id | 2327 | 2327 | int64 | |
| 1 | retweet_count | 2327 | 1671 | int64 | |
| 2 | favorite_count | 2327 | 2006 | int64 | |

Source: visual and programmatic assessment

```python
# #, Variable, Non-Null (Count), Dtype:
dfa.info()
# Nunique:
dfa.nunique()
```

In [74]:
dft.nunique()

tweet_id          2327
retweet_count     1671
favorite_count    2006
dtype: int64

### Quality issues
1.

2.

3.

4.

5.

6.

7.

8.

### Tidiness issues
1.

2.

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [ ]:
# Make copies of original pieces of data


### Issue #1:

#### Define:

#### Code

#### Test

### Issue #2:

#### Define

#### Code

#### Test

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1.

2.

3.

### Visualization